## Assignments
In the following exercises, you'll be working with the code snippet below:

In [1]:
#%%timeit
#x = da.random.random((10000, 10000), chunks=(1000, 1000))
#y = x + x.T
#z = y[::2, 5000:].mean(axis=1)
#z.compute()

In [3]:
import warnings
warnings.filterwarnings("ignore")

from dask.distributed import Client, progress

client = Client(n_workers=4, threads_per_worker=2, memory_limit='2GB')
client

<Client: 'tcp://127.0.0.1:60515' processes=4 threads=8, memory=8.00 GB>

In [9]:
import dask.array as da
#import numpy as np

In [5]:
##

In [10]:
%%timeit
x = da.random.random((10000, 10000), chunks=(1000, 1000))
y = x + x.T
z = y[::2, 5000:].mean(axis=1)
z.compute()

872 ms ± 116 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Took around 872 milliseconds per loop.

To complete this assignment, submit your solutions to the following tasks as a link to your Jupyter Notebook on GitHub.

1. Change the code above by setting chunks=(250, 250). How long does it take to run?
2. Now, set the parameter to chunks=(500, 500). How long does it take to run? Does this one or the previous one run more quickly? Why?


Submit your work below. You can also take a look at this example solution.

https://drive.google.com/file/d/1ulyVmcIbFaYFd6_zQO4UzLkKVWFM5BsP/view?usp=sharing

1. Change the code above by setting chunks=(250, 250). How long does it take to run?

In [11]:
%%timeit
x = da.random.random((10000, 10000), chunks=(250, 250))
y = x + x.T
z = y[::2, 5000:].mean(axis=1)
z.compute()

2.99 s ± 160 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Took around 2.99 seconds/2990 milliseconds per loop.

2. Now, set the parameter to chunks=(500, 500). How long does it take to run? Does this one or the previous one run more quickly? Why?

In [12]:
%%timeit
x = da.random.random((10000, 10000), chunks=(500, 500))
y = x + x.T
z = y[::2, 5000:].mean(axis=1)
z.compute()

1.37 s ± 128 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Took around 1.37 seconds/1370 milliseconds per loop.

Chunks of 1000 ran the fastest, then chunks of 500. The slowest was chunks of 250. 

Chunks are what the whole dataframe is split into. If there are smaller chunks, then more runs have to occur. 

Took about half as long in 500 chunks than in 250 chunks. 

When we divide a Dask array into smaller chunks Dask needs to manage a higher number of chunks. This requires more computation. If our dataset can fit into memory as is in our case, then setting the chunk sizer to a larger number speeds up the computation.

In [13]:
client.close()